google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,389 @@
+      builder.setExecutorProvider(executorProvider);
+    }
+    if (endpoint != null) {
+      builder.setEndpoint(endpoint);
+    }
+    JsonStreamWriterOnSchemaUpdateRunnable jsonStreamWriterOnSchemaUpdateRunnable =
+        new JsonStreamWriterOnSchemaUpdateRunnable();
+    jsonStreamWriterOnSchemaUpdateRunnable.setJsonStreamWriter(this);
+    builder.setOnSchemaUpdateRunnable(jsonStreamWriterOnSchemaUpdateRunnable);
+  }
+
+  /**
+   * Setter for descriptor.
+   *
+   * @param tableSchema
+   */
+  public synchronized void setDescriptor(Table.TableSchema tableSchema)
+      throws Descriptors.DescriptorValidationException {
+    this.descriptor =
+        BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(tableSchema);
+  }
+
+  /**
+   * newBuilder that constructs a JsonStreamWriter builder with BigQuery client being initialized by
+   * StreamWriter by default.
+   *
+   * @param streamName name of the stream that must follow
+   *     "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+"
+   * @param tableSchema The schema of the table when the stream was created, which is passed back
+   *     through {@code WriteStream}
+   * @return Builder
+   */
+  public static Builder newBuilder(String streamName, Table.TableSchema tableSchema) {
+    Preconditions.checkNotNull(streamName, "StreamName is null.");
+    Preconditions.checkNotNull(tableSchema, "TableSchema is null.");
+    return new Builder(streamName, tableSchema, null);
+  }
+
+  /**
+   * newBuilder that constructs a JsonStreamWriter builder.
+   *
+   * @param streamName name of the stream that must follow
+   *     "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+"
+   * @param tableSchema The schema of the table when the stream was created, which is passed back
+   *     through {@code WriteStream}
+   * @param client
+   * @return Builder
+   */
+  public static Builder newBuilder(
+      String streamName, Table.TableSchema tableSchema, BigQueryWriteClient client) {
+    Preconditions.checkNotNull(streamName, "StreamName is null.");
+    Preconditions.checkNotNull(tableSchema, "TableSchema is null.");
+    Preconditions.checkNotNull(client, "BigQuery client is null.");
+    return new Builder(streamName, tableSchema, client);
+  }
+
+  /** Closes the underlying StreamWriter. */
+  @Override
+  public void close() {
+    this.streamWriter.close();
+  }
+
+  public class JsonStreamWriterOnSchemaUpdateRunnable extends OnSchemaUpdateRunnable {
+    private JsonStreamWriter jsonStreamWriter;
+    /**
+     * Setter for the jsonStreamWriter
+     *
+     * @param jsonStreamWriter
+     */
+    public void setJsonStreamWriter(JsonStreamWriter jsonStreamWriter) {
+      this.jsonStreamWriter = jsonStreamWriter;
+    }
+
+    /** Getter for the jsonStreamWriter */
+    public JsonStreamWriter getJsonStreamWriter() {
+      return this.jsonStreamWriter;
+    }
+
+    @Override
+    public void run() {
+      this.setAttachUpdatedTableSchema(true);
```

Comment:
Instead of calling refreshAppend on streamwriter, could we have a method on JsonWriter that does refreshAppend and setDescriptor, so that we can lock the two operations together (with a lock that will be called by Append as well).

Reply:
We will still need locking inside StreamWriter since it is different layer of interface.

Reply:
Fixed, added a method called refreshConnection that calls refreshAppend, writeAllOutstanding, and setDescriptor.